tool/optimize_parser.rb

```diff
@@ -62,13 +62,23 @@ def generate_ripper_action_body_method(state, code_body)
     @out.puts "};"
   end
 
+  def action_body_method(code_body)
+    if code_body.empty?
+      # If the code body is empty, we can generate a Java lambda expression rather than the more verbose lambda statement.
+      'yyVal;'
+    else
+      # Generate a full Java lambda expression for the action.
+      ret = "{\n"
+
+      code_body.each { |line| ret << line }
+
+      ret << "    return yyVal;\n"
+      ret << "};\n"
+    end
+  end
+
   def generate_action_body_method(state, code_body)
-    @out.puts "states[#{state}] = new ParserState() {"
-    @out.puts "  @Override public Object execute(ParserSupport support, RubyLexer lexer, Object yyVal, Object[] yyVals, int yyTop) {"
-    code_body.each { |line| @out.puts line }
-    @out.puts "    return yyVal;"
-    @out.puts "  }"
-    @out.puts "};"
+    @out.puts "states[#{state}] = (support, lexer, yyVal, yyVals, yyTop) -> #{action_body_method(code_body)}"
   end
```

Comment:
Does this have any impact on performance? Especially startup time? It's a huge number of lambdas.

Reply:
I haven't noticed anything, but I'll spend time measuring. For posterity, it's 645 lambdas.

Reply:
To measure on startup, you can use `Launcher.printTruffleTimeMetric("before-parse") + "after-parse"` in `TranslatorDriver` + `jt metrics time`.
For warmed parser perf, I guess parsing a few 100 times the core lib would be a good workload.

Reply:
There's no significant time difference. I do see the lambda-based approach consistently requires upwards of ~0.1% more instructions according to `perf stat`.

Reply:
Is that on JVM?
On SVM the startup part shouldn't matter (but still good to check). Warmed parsing perf might change though.

Reply:
Sorry, yes, SVM. We still parse for things like RubyGems. Trying to use `perf stat` on the JVM is too chaotic.

Reply:
It's hard to really see on the JVM. It looks like the trend is for the non-lambda one to be a bit faster, but repeated measurements are ± 100ms.

I can back that bit out if it's concerning. I thought it made reading the parser code a bit simpler, but I doubt anyone spends a significant amount of time in there.

Reply:
> Sorry, yes, SVM. We still parse for things like RubyGems there.

Really not much code with lazy-rubygems.

> It's hard to really see on the JVM. It looks like the trend is for the non-lambda one to be a bit faster, but repeated measurements are ± 100ms. For the native image, it was closer to ± 10ms.

Yes, it's a bit more noisy.
But what do you get with `jt metrics time -e 0` for `load-core`? Isn't that stable enough?
Maybe using more precise metrics like I proposed above would help with getting good results.

> I can back that bit out if it's concerning. I thought it made reading the parser code a bit simpler, but I doubt anyone spends a significant amount of time in there.

It would be nice to have it in its own commit, but it's probably easy to revert anyway if we ever need that. I think it's OK here, but since it's a large change in the parser, it would be good to have a good idea on the effect of performance on both startup/warmed perf for JVM/SVM.

Reply:
Based on `jt metrics` I can't see any measurement difference I can't attribute to natural variance without rigorous statistical analysis. I'll push ahead as is. If anyone else sees otherwise, it's a pretty easy change to back out. The parser diff looks large, but it's generated by a very small amount of Ruby.

Reply:
Per @eregon's request, I modified `TranslatorDriver` to measure just the lexer setup and parsing, 100 times per file, to ensure things were reasonably warmed up. There was essentially no difference. I took 3 samples of each and the mean of each approach was within 10ms of each other.

Reply:
I ran through the same process with just a single parse pass for each file to measure cold time the mean was within 4ms of each other.

Reply:
Great, thank you for checking!